cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaNodeUnregistration.java

```diff
@@ -0,0 +1,99 @@
+     *
+     * @param reconciliation        Reconciliation marker
+     * @param vertx                 Vert.x instance
+     * @param adminClientProvider   Kafka Admin API client provider
+     * @param pemTrustSet           Trust set for the admin client to connect to the Kafka cluster
+     * @param pemAuthIdentity       Key set  for the admin client to connect to the Kafka cluster
+     * @param nodeIdsToUnregister   List of node IDs that should be unregistered
+     *
+     * @return  Future that completes when all nodes are unregistered
+     */
+    public static Future<Void> unregisterNodes(
+            Reconciliation reconciliation,
+            Vertx vertx,
+            AdminClientProvider adminClientProvider,
+            PemTrustSet pemTrustSet,
+            PemAuthIdentity pemAuthIdentity,
+            List<Integer> nodeIdsToUnregister
+    ) {
+        try {
+            String bootstrapHostname = KafkaResources.bootstrapServiceName(reconciliation.name()) + "." + reconciliation.namespace() + ".svc:" + KafkaCluster.REPLICATION_PORT;
+            Admin adminClient = adminClientProvider.createAdminClient(bootstrapHostname, pemTrustSet, pemAuthIdentity);
+
+            List<Future<Void>> futures = new ArrayList<>();
+            for (Integer nodeId : nodeIdsToUnregister) {
+                futures.add(unregisterNode(reconciliation, vertx, adminClient, nodeId));
+            }
+
+            return Future.all(futures)
+                    .eventually(() -> {
+                        adminClient.close();
+                        return Future.succeededFuture();
+                    })
+                    .map((Void) null);
+        } catch (KafkaException e) {
+            LOGGER.warnCr(reconciliation, "Failed to unregister nodes", e);
+            return Future.failedFuture(e);
+        }
+    }
+
+    /**
+     * Unregisters a single Kafka node using the Kafka Admin API. In case the failure is caused by the node not being
+     * registered, the error will be ignored.
+     *
+     * @param reconciliation        Reconciliation marker
+     * @param vertx                 Vert.x instance
+     * @param adminClient           Kafka Admin API client instance
+     * @param nodeIdToUnregister    ID of the node that should be unregistered
+     *
+     * @return  Future that completes when the node is unregistered
+     */
+    private static Future<Void> unregisterNode(Reconciliation reconciliation, Vertx vertx, Admin adminClient, Integer nodeIdToUnregister) {
+        LOGGER.debugCr(reconciliation, "Unregistering node {} from the Kafka cluster", nodeIdToUnregister);
+
+        return VertxUtil
+                .kafkaFutureToVertxFuture(reconciliation, vertx, adminClient.unregisterBroker(nodeIdToUnregister).all())
+                .recover(t -> {
+                    if (t instanceof BrokerIdNotRegisteredException)    {
+                        // The broker is not registered anymore, we report success
```

Comment:
what does exactly it mean? That the node was already unregistered from the cluster (how?) but we are still trying to unregister it? (so the exception?). Can this happen?

Reply:
Another situation where this could happen is if the operator fails before updating the status in the Kube cluster etc.

Reply:
Makes sense. Thanks.

Reply:
I will extend the comment in the code to make it more clear from there.